### PR TITLE
Integrate blockchain component and Pyrsia p2p client

### PR DESCRIPTION
**Pull request by fishseabowl**
## Description

Fixes pyrsia/pyrsia#977

This PR add BlockchainService module to integrate blockchain component and Pyrsia p2p client to support updating blocks between different nodes. 

## PR Checklist

- [x] I've read the [contributing guidelines](https://github.com/pyrsia/.github/blob/main/contributing.md).
- [x] I've read ["What is a Good PR?"](https://github.com/pyrsia/pyrsia/blob/main/docs/get_involved/good_pr.md)
- [x] I've included a good title and brief description along with how to review them.
- [x] I've linked any associated an [issue](https://github.com/pyrsia/pyrsia/issues).

### Code Contributions

- [x] I've built the code `cargo build --all-targets` successfully.
- [x] I've run the unit tests `cargo test --workspace` and everything passes.
- [x] I've made sure my rust toolchain is current `rustup update`.

### CODE REVIEW

**Comment by betarelease** [comment]
This is quite a big PR and I am not able to understand it well. Can you do a demo to the team during standup so that we can only comment on parts that are critical. @fishseabowl

**Comment by betarelease** [comment]
Can @erwin1 and @karolh2000 review this so that we can merge this soon? This has been here for a while now.

**Comment by erwin1** [comment]
@fishseabowl the only reservation I still have is the drop in codecoverage. But maybe we can't do anything about it now?

**Comment by fishseabowl** [comment]
> @fishseabowl the only reservation I still have is the drop in codecoverage. But maybe we can't do anything about it now?

I think I can add more error handling to improve code coverage.  Thanks

File: src/blockchain_service/service.rs [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+    }
+
+    /// Add payload to blockchain
+    pub async fn add_payload(&mut self, payload: Vec<u8>, local_key: &identity::Keypair) {
+        let _ = self.blockchain.add_block(payload, local_key).await;
+        self.notify_block_update(Box::new(self.blockchain.last_block().unwrap()))
+            .await;
+    }
+
+    pub async fn notify_block_update(&mut self, block: Box<Block>) {
```
Comment (tiainen) [inline src/blockchain_service/service.rs:54]: This function can be private?

File: src/blockchain_service/service.rs [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+impl BlockchainService {
+    pub fn new(keypair: &identity::ed25519::Keypair, p2p_client: Client) -> Self {
+        Self {
+            blockchain: Blockchain::new(keypair),
+            p2p_client,
+        }
+    }
+
+    /// Add payload to blockchain
+    pub async fn add_payload(&mut self, payload: Vec<u8>, local_key: &identity::Keypair) {
```
Comment (tiainen) [inline src/blockchain_service/service.rs:48]: This function does not seem to be used anywhere?

File: src/blockchain_service/service.rs [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+            .await;
+    }
+
+    pub async fn notify_block_update(&mut self, block: Box<Block>) {
+        let peer_list = self.p2p_client.list_peers().await.unwrap_or_default();
+        let block_ordinal = block.header.ordinal;
+        for peer_id in peer_list.iter() {
+            let _ = self
+                .p2p_client
+                .request_block_update(peer_id, block_ordinal, block.clone())
```
Comment (tiainen) [inline src/blockchain_service/service.rs:60]: I was wondering why you need to send the ordinal separately when it's taken from the header of the block?

File: src/blockchain_service/service.rs (hunk 1 above)
Comment (fishseabowl) [reply src/blockchain_service/service.rs:54]: Yes. Currently this function is only used by blockchain services. Maybe private is better.

File: src/blockchain_service/service.rs (hunk 2 above)
Comment (fishseabowl) [reply src/blockchain_service/service.rs:48]: The transparent logging service uses this function to notify the blockchain service about new payloads.

File: src/blockchain_service/service.rs (hunk 3 above)
Comment (fishseabowl) [reply src/blockchain_service/service.rs:60]: It improves performance without decoding the block structure to determine some invalid block.

**Review by erwin1 — APPROVED** [review]

**Review by tiainen — APPROVED** [review]

**Pull request merged**